lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1308,7 +1308,24 @@ void TypeChecker::checkIgnoredExpr(Expr *E) {
   // dead?
   if (E->getType()->is<AnyFunctionType>()) {
     bool isDiscardable = false;
-    if (auto *Fn = dyn_cast<ApplyExpr>(E)) {
+
+    // The called function could be wrapped inside a `dot_syntax_call_expr`
+    // node, for example:
+    //
+    // class Bar {
+    //   @discardableResult
+    //   func foo() -> Int { return 0 }
+    //
+    //   func baz() {
+    //     self.foo
+    //   }
+    // }
+    //
+    // So look through the DSCE and get the function being called.
+    auto expr =
+        isa<DotSyntaxCallExpr>(E) ? dyn_cast<DotSyntaxCallExpr>(E)->getFn() : E;
```

Comment:
Nitpick: You can use `cast` instead of `dyn_cast` since you just checked it with `isa`.

Reply:
Done @jrose-apple 👍